### ISSUE TITLE
[Question] How use intra_var_miner for multiclass

### ISSUE THREAD

**Issue by cfrancois7** [KevinMusgrave/pytorch-metric-learning#378]
Firstly, thank you for your great package and modules.
It works very fine, well documented.

Nevertheless, I struggle to determine what kind of loss I can use for binary or multiclass.
For example, I'm trying to create a kind of "anomaly" detector through metric learning.
In the ideal world, in the embedding space I would like "correct" events as cluster and the anomaly for from them.

I have 3 classes :
- class A,
- class B and
- class other/anomaly.
2 kinds of correct event (but different, and I want discriminate them) and many anomalies, I tried to define a multi loss in this way and I got good results but not perfect.

Ideally, in the space embedding, I just want 2 clusters (A and B) but no constrains about other/anomaly because it could be the rest of the world.
So I tried to used IntraPairVarianceLoss only on positives. Not sure it works here. In binary classification, I understand "pos_margin" and "neg_margin" for the miner, but when we want the constraint only on some classes (and not all of them)??
Also, is it very coherent to use TripletMargin with Arcface? I works quite fine, but I'm not confident.

I'm looking for some "intuition" about how to use loss that I understand very well for binary classification (and works very well) but for multiclass.

```
# MINERS
miner = miners.TripletMarginMiner(
    margin=triplet_miner_margin,
    type_of_triplets='all',
    distance = CosineSimilarity(),
    ) # maybe not relevant in multiclass, but seems to work fine
    
intra_var_miner = miners.PairMarginMiner(
    pos_margin=1.,
    neg_margin=1.,
    distance=CosineSimilarity()) # I would like a minimum of intra_variance but only for my "correct" class.

angular_miner = miners.AngularMiner(angle=20) # works well in multiclass.

# REDUCER
reducer = reducers.AvgNonZeroReducer()

# LOSSES
var_loss = losses.IntraPairVarianceLoss(reducer=reducer)
arcface_loss = losses.ArcFaceLoss(3, embedding_size, margin=28.6, scale=64, reducer=reducer)
main_loss = losses.TripletMarginLoss( 
    margin=triplet_loss_margin,
    distance = CosineSimilarity(),
    embedding_regularizer = regularizers.LpRegularizer(),
    reducer=reducer,
)
loss_func = losses.MultipleLosses(
    [main_loss, var_loss, arcface_loss],
    miners=[miner, intra_var_miner, None],
    # weights=multiple_loss_weights
)
```

**Comment by KevinMusgrave** [comment]
> I have 3 classes :
> 
> * class A,
> * class B and
> * class other/anomaly.
>   2 kinds of correct event (but different, and I want discriminate them) and many anomalies, I tried to define a multi loss in this way and I got good results but not perfect.
> 
> Ideally, in the space embedding, I just want 2 clusters (A and B) but no constrains about other/anomaly because it could be the rest of the world. So I tried to used IntraPairVarianceLoss only on positives. Not sure it works here. In binary classification, I understand "pos_margin" and "neg_margin" for the miner, but when we want the constraint only on some classes (and not all of them)?? 

Do you mean you want to apply the loss only to samples with labels A and B?

>Also, is it very coherent to use TripletMargin with Arcface? I works quite fine, but I'm not confident.

TripletMarginMiner with ArcFace is fine. The triplets will be converted to weights, i.e. each sample in the batch will be weighted by how many times it appears in the mined triplets.

> I'm looking for some "intuition" about how to use loss that I understand very well for binary classification (and works very well) but for multiclass.

By "multiclass" do you mean there are more than 2 classes and each sample can only have 1 label, or do you mean that each sample can have multiple labels?

**Comment by cfrancois7** [comment]
> Do you mean you want to apply the loss only to samples with labels A and B?

Yes. It is exaclty what I'm looking for.
I tried PairMarginMiner but it does not what I'm looking for.

**Why I'm trying this ?**
In binary classification (class A vs other/anomaly) I noted that I got better results when the `IntraPairVarianceLoss` is applied only to `class A`. The hypothesis behind this is the class `other/anomaly` has high variety and trying to build a cluster from them loosens the convergence.

Should develop something adhoc? Should I used two IntraPairVarianceLoss? One for each task ?

Multiclass is several categories. Here I've only one label by sample and one label corresponding to one class.

**Comment by KevinMusgrave** [comment]
This should apply the loss to positive pairs of (1,1) and (2,2)

```python
from pytorch_metric_learning.utils import loss_and_miner_utils as lmu

not_anomaly = labels != 0
emb_, labels_ = embeddings[not_anomaly], labels[not_anomaly]
a1, p, _, _ = lmu.get_all_pairs_indices(labels_)
a2, n = torch.tensor([]), torch.tensor([])
loss = loss_fn(emb_, labels_, (a1,p,a2,n))
```

**Comment by cfrancois7** [comment]
Thank you a lot. It works. However it double the loss calculation (each pair is selected twice).
I decided to create a class to deal with this : 
```
       a1 = torch.where(not_anomaly)[0][a1]
        p = torch.where(not_anomaly)[0][p]
        pos_pair = mat.triu()[a1, p]
```

```
class PositiveIntraVarMiner(PairMarginMiner):
    """
    Returns only positive pairs that have distance greater than a margin
    """

    def __init__(self, pos_margin=0.2, **kwargs):
        super().__init__(**kwargs)
        self.pos_margin = pos_margin
        self.add_to_recordable_attributes(list_of_names=["pos_margin"], is_stat=False)

    def mine(self, embeddings, labels, ref_emb, ref_labels):
        not_anomaly = labels != 0
        _labels = labels[not_anomaly]
        a1, p, _, _ = lmu.get_all_pairs_indices(_labels)
        mat = self.distance(embeddings, ref_emb)
        a1 = torch.where(not_anomaly)[0][a1]
        p = torch.where(not_anomaly)[0][p]
        pos_pair = mat.triu()[a1, p]
        self.set_stats(pos_pair)
        pos_mask = (
            pos_pair < self.pos_margin
            if self.distance.is_inverted
            else pos_pair > self.pos_margin
        )
        a2, n = torch.tensor([]), torch.tensor([])
        return a1[pos_mask], p[pos_mask], a2, n

    def set_stats(self, pos_pair):
        if self.collect_stats:
            with torch.no_grad():
                self.pos_pair_dist = (
                    torch.mean(pos_pair).item() if len(pos_pair) > 0 else 0
                )
```

**Comment by KevinMusgrave** [comment]
Looks good
